ffi/harfbuzz/src/lib.rs

```diff
@@ -69,606 +73,372 @@ use icu_properties::{GeneralCategory, Script};
-    ///
-    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
-    ///
-    /// [📚 Help choosing a constructor](icu_provider::constructors)
-    #[cfg(feature = "compiled_data")]
-    pub fn new() -> Result<Self, HarfBuzzError> {
-        let ufuncs = Self::empty()?;
-
-        unsafe {
-            hb_unicode_funcs_set_combining_class_func(
-                ufuncs.raw,
-                Some({
-                    extern "C" fn cb(
-                        _: *mut hb_unicode_funcs_t,
-                        unicode: hb_codepoint_t,
-                        _: *mut c_void,
-                    ) -> hb_unicode_combining_class_t {
-                        CanonicalCombiningClassMap::new().get32(unicode).0
-                            as hb_unicode_combining_class_t
-                    }
-                    cb
-                }),
-                core::ptr::null_mut(),
-                None,
-            );
-        }
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(ANY, Self::try_new_unstable)]
+    pub fn try_new_with_any_provider(
+        provider: &(impl icu_provider::AnyProvider + ?Sized),
+    ) -> Result<Self, HarfBuzzError> {
+        Self::try_new_unstable(&provider.as_downcasting())
+    }
+    #[cfg(feature = "serde")]
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(BUFFER,Self::try_new_unstable)]
+    pub fn try_new_with_buffer_provider(
+        provider: &(impl icu_provider::BufferProvider + ?Sized),
+    ) -> Result<Self, HarfBuzzError> {
+        Self::try_new_unstable(&provider.as_deserializing())
+    }
+}
 
-        unsafe {
-            hb_unicode_funcs_set_general_category_func(
-                ufuncs.raw,
-                Some({
-                    extern "C" fn cb(
-                        _: *mut hb_unicode_funcs_t,
-                        unicode: hb_codepoint_t,
-                        _: *mut c_void,
-                    ) -> hb_unicode_general_category_t {
-                        // Indexing is OK, because `GeneralCategory` data is validated upon
-                        // deserialization so that there can be no out-of-range `GeneralCategory`
-                        // values (which would be UB to materialize). `GeneralCategory` is a
-                        // stable exhaustive enum, and the length of `ICU4X_GENERAL_CATEGORY_TO_HARFBUZZ`
-                        // matches the number of enum items, so the index will always be in range here.
-                        #![allow(clippy::indexing_slicing)]
-                        ICU4X_GENERAL_CATEGORY_TO_HARFBUZZ
-                            [maps::general_category().get32(unicode) as usize]
-                            as hb_unicode_general_category_t
-                    }
-                    cb
-                }),
-                core::ptr::null_mut(),
-                None,
-            );
+fn convert_gc(gc: GeneralCategory) -> harfbuzz_traits::GeneralCategory {
+    match gc {
```

Comment:
Observation: ICU4C and ICU4X agree on the numbering, but Harfbuzz uses different numbering.

https://github.com/unicode-org/icu/blob/main/icu4c/source/common/unicode/uchar.h#L820

https://github.com/servo/rust-harfbuzz/blob/main/harfbuzz-traits/src/unicode_funcs_traits.rs#L13

https://github.com/unicode-org/icu4x/blob/main/components/properties/src/props.rs#L831

Reply:
It usually generates a jump table, which does typically work that way.

Reply:
This strikes me as a premature optimization but you're welcome to make it as a followup.